bin/cohttp_server_lwt.ml

```diff
@@ -81,58 +64,31 @@ let handler ~info ~docroot ~verbose ~index (ch,conn) req body =
         >>= Lwt_list.map_s (fun f ->
           let file_name = file_name / f in
           Lwt.try_bind
-            (fun () -> Lwt_unix.stat file_name)
-            (fun stat -> return (Some stat.Unix.st_kind, f))
-            (fun exn -> return (None, f)))
+            (fun () -> Lwt_unix.LargeFile.stat file_name)
+            (fun stat ->
+              return (Some (kind_of_unix_kind stat.Unix.LargeFile.st_kind),
+                      stat.Unix.LargeFile.st_size,
+                      f))
+            (fun exn -> return (None, 0L, f)))
         >>= fun listing ->
-        let html = List.map (fun (kind, f) ->
-          let encoded_f = Uri.pct_encode f in
-          match kind with
-          | Some Unix.S_DIR ->
-            let link = Uri.with_path uri (path / encoded_f / "") in
-            li link (sprintf "<i>%s/</i>" f)
-          | Some Unix.S_REG ->
-            let link = Uri.with_path uri (path / encoded_f) in
-            li link f
-          | Some _ -> sprintf "<li><s>%s</s></li>" f
-          | None -> sprintf "<li>Error with file: %s</li>" f
-        ) (sort listing) in
-        let contents = String.concat "\n" html in
-        let body = sprintf "
-              <html>
-                <body>
-                <h2>Directory Listing for <em>%s</em></h2>
-                <ul>%s</ul>
-                <hr />%s
-                </body>
-              </html>"
-          (Uri.pct_decode path) contents info in
+        let body = html_of_listing uri path (sort listing) info in
         Server.respond_string ~status:`OK ~body ()
     end
-    | Unix.S_REG -> serve_file ~docroot ~uri
+    | `File -> serve_file ~docroot ~uri
     | _ ->
       Server.respond_string ~status:`Forbidden
-        ~body:(sprintf "<html><body><h2>Forbidden</h2>
-        <p><b>%s</b> is not a normal file or directory</p>
-        <hr />%s</body></html>" path info)
+        ~body:(html_of_forbidden_unnormal path info)
         ()
   ) (function
   | Unix.Unix_error(Unix.ENOENT, "stat", p) as e ->
     if p = file_name
     then Server.respond_string ~status:`Not_found
-      ~body:(sprintf "<html><body><h2>Not Found</h2>
-      <p><b>%s</b> was not found on this server</p>
-      <hr />%s</body></html>" path info)
+      ~body:(html_of_not_found path info)
       ()
     else fail e
   | e -> fail e
   )
 
-let string_of_sockaddr = function
```

Comment:
Yes, unused but perhaps should be used somewhere. The async and lwt servers still don't have equivalent console logging.